### ISSUE TITLE
Various errors (fatal lock manager error, pthread_mutex_destroy failed) caused by races when opening/closing database

### ISSUE THREAD

**Issue by fperana** [FirebirdSQL/firebird#7371]
We've recently switched our production server to FB 4.0.2 from FB 3.0.7, both as SuperServer.

**Current OS is Debian 11 with 4 CPU and 24GB of RAM running on VMWare.**
Previous OS was CentOS 7, same HW config.

We have about 20-25 DBs accessed through this server by an average of 30-40 concurrent remote connections via TCP/IP (the clients are Delphi applications using FireDAC).
Some of these connections are "steady" (the application connects to the DB and stays there until it's closed), some others are short lived ones from a stateless application server, so maybe 15-25 connections are steady ones, the other 10-15 are short lived ones.

With FB 3 we had absolutely no problems, everything worked fine for years.

**Now with FB 4.0.2 we're experiencing some serious problems (engine crashes)**, luckily not so often, but when it happes all the connections and active transactions obviously are gone.

In a week we had 5 of these events and they're almost all different, but I suspect there's a common cause to them.

This is what I see in the firebird.log (I kept only the error lines):

```
Fatal lock manager error: invalid lock id (94352), errno: 0
Operating system call pthread_mutex_destroy failed. Error code 16
Fatal lock manager error: Process disappeared in LockManager::acquire_shmem, errno: 0
Operating system call pthread_mutex_destroy failed. Error code 16
Operating system call pthread_mutex_lock failed. Error code 22
```

All these errors led to a restart of the engine by the guardian:

```
/opt/firebird/bin/fbguard: /opt/firebird/bin/firebird terminated abnormally (-1)
/opt/firebird/bin/fbguard: guardian starting /opt/firebird/bin/firebird

```
Have you got an idea what could be the cause?

Unfortunately we've changed both FB version and OS, so I can't say which of them is the root of these problems... Maybe there's some config I can change to mitigate or eliminate these problems, or maybe it's a FB bug?

Thanks!

**Comment by AlexPeshkoff** [comment]
First of all I need full log, not extracts from it. Next - tune your system to provide core files in a case of crashes, install debuginfo and send me stack traces for a moment of the crash (for details see http://www.ibphoenix.com/resources/documents/search/doc_36). 
Except core dump there will probably be file fb_lock_table.dump. Save it together with core dump.

**Comment by fperana** [comment]
I've extracted the error lines from firebird.log just because the other lines are OK. I'll file the full log with the other infos.
Just a question: is it safe to put a debug build in production?

**Comment by fperana** [comment]
@AlexPeshkoff the link you provided seems a bit outdated, all major linux distros use systemd nowadays, so what I did is:
- Set `BugcheckAbort = 1` in `firebird.conf`
- Added `LimitCORE=infinity` under `[Service]` section in `firebird.service` systemd unit

Is that enough? If yes, I'll wait for the next crash and continue with the debuginfo thing.

I'm not sure if the `LimitCORE` directive will affect the main firebird process, given that systemd is starting the guardian, not the main engine... I hope this directive will "propagate" to that child process.

**Comment by AlexPeshkoff** [comment]
You should distinguish between debug build and debuginfo for release build. Debug build is usually used by developers, it works significantly slower than release one and should not be used in production. Running release build with debuginfo is absolutely safe - it affects nothing before you use gdb.

And yes - LimitCORE propagates to child.

BTW, I need to know - what is the source of your FB binaries?

**Comment by fperana** [comment]
> You should distinguish between debug build and debuginfo for release build.

Yes, I had figured that out after reading your linked page ;)
During this evening I'll restart the firebird service, so both LimitCORE and BugcheckAbort will be active.

> BTW, I need to know - what is the source of your FB binaries?

I'm using the official build from https://firebirdsql.org/en/firebird-4-0, specifically the [Linux AMD64](https://github.com/FirebirdSQL/firebird/releases/download/v4.0.2/Firebird-4.0.2.2816-0.amd64.tar.gz) one.

**Comment by fperana** [comment]
I'm here with a core dump, but first a little explanation.

I were suspecting that the cause could be this config in the firebird.conf: `StatementTimeout = 600`

This is the only different configuration I have on the production server than on the test one. So I've tried to comment it out and there were no more crashes. I've also commented out the `ConnectionIdleTimeout = 30` config, just to be on the safe side, but this was already there on the test server for more than a year now and it never crashed...

Anyway, after 10 days of fine service, I've re-added these two config lines and after some hours we got a crash.

Here is the [firebird.log](https://github.com/FirebirdSQL/firebird/files/10004528/firebird.log)
(again, it's not the whole one, but the lines before and after the crash are just repeating over and over with similar frequency, this is caused by a particular client that is being killed by a supervisor, giving it no chances to disconnect the right way - we're working to fix this, but this situation was working fine on FB3 for years and on FB4 without `StatementTimeout` configuration, so I'm confident this is not related).

This is the [stack_trace_firebird.txt](https://github.com/FirebirdSQL/firebird/files/10004529/stack_trace_firebird.txt)
(I hope to have captured it the right way, I see some warnings/errors at the start, but I have no idea what is going wrong or if it's ok)

Lastly, I found no `fb_lock_table.dump` file on the whole filesystem...

Hope this can help tracking down the problem.

**Comment by AlexPeshkoff** [comment]
That's other error, not related directly with lock manager.

BTW, You can avoid the worst warning (core file may not match specified executable file) starting gdb this way:
# gdb /opt/firebird/bin/firebird core
(debug info will be picked automatically)

**Comment by fperana** [comment]
> That's other error, not related directly with lock manager.

Sorry, maybe I don't understand what you mean exactly, but this is one of the four different kind of errors we've experienced the first week we got FB 4.0.2 running (as you can see in my first post) with `StatementTimeout = 600` config.
As you can see, a couple were **lock manager** related, the other three are **OS pthread_mutex** errors. Not the same thing, but I suspect they're related.

**Comment by AlexPeshkoff** [comment]
Is not it too late to ask for core dump for last error? If you have it - please also provide some libraries (no symlinks!) from you host:
libc.so.6  libgcc_s.so.1  libm.so.6  libpthread.so.0  libstdc++.so.6  libthread_db.so.1

**Comment by fperana** [comment]
I still have it. Here's a zip with the core dump and the libraries you asked.

[fb402_core_libs.zip](https://github.com/FirebirdSQL/firebird/files/10066348/fb402_core_libs.zip)

Apart from `libgcc_s.so.1` which is a real file itself, all the other were symlinks. Here's a list of the real filenames, if it's somewhat important for you:
`libc.so.6 -> libc-2.31.so`
`libm.so.6 -> libm-2.31.so`
`libpthread.so.0 -> libpthread-2.31.so`
`libstdc++.so.6 -> libstdc++.so.6.0.28`
`libthread_db.so.1 -> libthread_db-1.0.so`

As a side note, I've run the system for 4 or 5 days with only `ConnectionIdleTimeout = 30` setting, and now it's running for another 4 days with only `StatementTimeout = 600`, there were no crashes during this period. I'll keep running with only one of the two settings for a couple of weeks, then switch again if no crashes will occur.

**Comment by AlexPeshkoff** [comment]
May be I was not enough specific - to analyze your core dump I need also mentioned dynamic libraries (libc-2.31.so, libpthread-2.31.so, etc.)

**Comment by fperana** [comment]
The libs are all in the zip file I've uploaded, with the names you mentioned (not the name of the real file). If you need them to have the original file name, please rename them accordingly to the list I given you (I should do the same and reupload...).

**Comment by fperana** [comment]
Maybe you're having issues decompressing the archive because I've compressed it using 7zip. It's in zip format but it uses lzma algorithm for better compression.
Here's a classic deflated zip file:
[fb402_core_libs.zip](https://github.com/FirebirdSQL/firebird/files/10068664/fb402_core_libs.zip)

**Comment by AlexPeshkoff** [comment]
Sorry - everything fine, I just did not understand you. And everything unpack ok.

**Comment by AlexPeshkoff** [comment]
I've added 2 related fixes - one should help avoid segfaults (https://github.com/FirebirdSQL/firebird/issues/7393), another w/o ticket will produce better core dumps in lock-manager related cases. I.e. if you can switch to current snapshot it can be useful.

**Comment by fperana** [comment]
This is a production server, a snapshot build may contain untested code, I don't think it's feasible for us to use it there.
I also have a test server, but there we never experienced these problems (it's a lot less "stressed" than the production one).

Anyway, during the last two weeks I've been playing around with `StatementTimeout` and `ConnectionIdleTimeout` settings (commenting and uncommenting them alternatively) and so far we experienced no more crashes. I suspect there may be some interference between the two, because the two times I enabled both, it's never gone more than 24-36 hours without crashing. I'll keep the current settings for a few more days, then if no crash will occur, I'll try re-enabling both to see if I can empirically confirm this.

**Comment by fperana** [comment]
Little update: today we experienced another crash, this time only `StatementTimeout` was enforced.
The error was the same of the last time: `Operating system call pthread_mutex_lock failed. Error code 22`, so again not directly related to the lock manager.
The core dump is much bigger than the previous one, about 1.4GB, compressed it's about 46MB, do you want me to upload it?

**Comment by AlexPeshkoff** [comment]
On 11/29/22 12:37, F. Perana wrote:
>
> Little update: today we experienced another crash, this time only 
> |StatementTimeout| was enforced.
> The error was the same of the last time: |Operating system call 
> pthread_mutex_lock failed. Error code 22|, so again not directly 
> related to the lock manager.
> The core dump is much bigger than the previous one, about 1.4GB, 
> compressed it's about 46MB, do you want me to upload it?
>

Yes please, will take a look at it.

**Comment by fperana** [comment]
Here is it. GitHub has a 25MB limit, so I've had to split the archive in 2 files, you need to merge them with some tool. And again, I've used 7zip to achieve better compression using LZMA algorithm.
And sorry for the file names, I forgot to rename them, Win32 is a subfolder of a delphi project 😛 

[Win32.001.zip](https://github.com/FirebirdSQL/firebird/files/10115033/Win32.001.zip)
[Win32.002.zip](https://github.com/FirebirdSQL/firebird/files/10115037/Win32.002.zip)

**Comment by fperana** [comment]
Oops, the core dump...
[core_20221130T102249.002.zip](https://github.com/FirebirdSQL/firebird/files/10121928/core_20221130T102249.002.zip)
[core_20221130T102249.001.zip](https://github.com/FirebirdSQL/firebird/files/10121931/core_20221130T102249.001.zip)

**Comment by fperana** [comment]
Another crash also today:

```
sqlx	Fri Dec  2 13:41:23 2022
	Operating system call pthread_mutex_lock failed. Error code 22

sqlx	Fri Dec  2 13:41:23 2022
	Operating system call pthread_mutex_lock failed. Error code 22

sqlx	Fri Dec  2 13:41:29 2022
	/opt/firebird/bin/fbguard: /opt/firebird/bin/firebird terminated abnormally (-1)
```
I've noticed that the `pthread_mutex_lock` error is always logged twice, while the other types of errors I've encountered so far are only logged once.

I'm trying to get as more infos as possible for you to understand and fix this issue. We're starting to have some serious problems if this is going to happen so often when the DB is under load.

**Comment by AlexPeshkoff** [comment]
On 12/2/22 16:47, F. Perana wrote:
>
> Another crash also today:
>
> |sqlx Fri Dec 2 13:41:23 2022 Operating system call pthread_mutex_lock 
> failed. Error code 22 sqlx Fri Dec 2 13:41:23 2022 Operating system 
> call pthread_mutex_lock failed. Error code 22 sqlx Fri Dec 2 13:41:29 
> 2022 /opt/firebird/bin/fbguard: /opt/firebird/bin/firebird terminated 
> abnormally (-1) |
>
> I've noticed that the |pthread_mutex_lock| error is always logged 
> twice, while the other types of errors I've encountered so far are 
> only logged once.
>
> I'm trying to get as more infos as possible for you to understand and 
> fix this issue. We're starting to have some serious problems if this 
> is going to happen so often when the DB is under load.
>

Can you run special build that will log additional info? Suppose this 
should help.

**Comment by fperana** [comment]
Can I run it in production with confidence it will not create worse problems?

Also, please give me a link to the snapshot I should install.

**Comment by AlexPeshkoff** [comment]
In fact we can avoid using snapshots. Please run (as root):
echo 0x3f >/proc/<firebird-PID>/coredump_filter
This will make next core dump bigger but more informative.

**Comment by fperana** [comment]
Today we experienced another crash, I have a core dump but unfortunately I forgot to remove a crontab script that restarts the firebird service during night (I did that to try to mitigate the crashes), so the 0x3f flag were restored to 0x33 on the new process :(
I've now removed the script and reissued the echo command, so we have to wait for the next crash.
BTW, I have seen that this way we activate the "dump file-backed private and shared mappings", would it be interesting to activate also the "Dump shared huge pages" flag? (0x7f)

**Comment by AlexPeshkoff** [comment]
On 12/6/22 12:27, F. Perana wrote:

> BTW, I have seen that this way we activate the "dump file-backed 
> private and shared mappings", would it be interesting to activate also 
> the "Dump shared huge pages" flag? (0x7f)
>

We do not use huge pages.

**Comment by fperana** [comment]
So today we're "lucky", just got another crash, this time 0x3f in place.
Here's the core dump, as usual zipped (lzma) and splitted. Hope this helps.
[core_20221206T154133.001.zip](https://github.com/FirebirdSQL/firebird/files/10167489/core_20221206T154133.001.zip)
[core_20221206T154133.002.zip](https://github.com/FirebirdSQL/firebird/files/10167493/core_20221206T154133.002.zip)
[core_20221206T154133.003.zip](https://github.com/FirebirdSQL/firebird/files/10167501/core_20221206T154133.003.zip)

**Comment by AlexPeshkoff** [comment]
Unfortunately that did not help. Afraid that w/o special build we can't 
move further. Will prepare it in a day or two and notify you. That will 
not be std snapshot - instead special build of 4.0.2 with some debugging 
facilities. I.e. minimum changes.

**Comment by AlexPeshkoff** [comment]
BTW, any specific usage patterns with 
/data/dbroot/scudonline/scudonline.fdb ?
Always fails when attaching it.

**Comment by fperana** [comment]
Thank you Alexander, I'll wait for it. I'll be in holidays from 8 dec to 11, so if I'll see something tomorrow I'll install during the night, else I'll try next monday.

**Comment by fperana** [comment]
> BTW, any specific usage patterns with /data/dbroot/scudonline/scudonline.fdb ? Always fails when attaching it.

Hmm, not that I'm aware of, but this is a starting point. I'll do some research tomorrow.

**Comment by AlexPeshkoff** [comment]
https://drive.google.com/drive/folders/1KXu5S215dN7dsthiAxnStZkttubRFgI-?usp=sharing

**Comment by fperana** [comment]
Thank you! I'd like to put this in service as soon as possible, but I cant stop my clients now for the whole time of an uninstall+reinstall+replace config files. Is it feasible to swap only one (or some) files from the new package, so I can stop firebird for just some seconds?

**Comment by AlexPeshkoff** [comment]
First notice - DB name now is sicurezza, i.e. bad things happen not only with scudonline.

**Comment by fperana** [comment]
Ok. In the meantime, I've investigated the usage patterns on scudonline and I haven't found anything in particular.

One thing that comes to my mind is that these two DBs are mostly affected by small and frequent connections/transactions.

**Comment by AlexPeshkoff** [comment]
On 12/13/22 14:36, F. Perana wrote:
> DBs are mostly affected by small and frequent connections

Yes, it fails always at attch time.

**Comment by AlexPeshkoff** [comment]
Now it's quite clear what happens, unfortunately no idea - why. (States of process and attached to it shared memory are incompatible).
Next, I've tried to reproduce your bug (a few opened DBs and a pair of isql's permanently doing close/open of database) locally. Latest binaries (exactly one I've sent to you) work fine in a loop of 65 536 iterations attach/detach.

I can add some logging to firebird in probably problematic places - but my suggestion to you is to try without vmware.

**Comment by fperana** [comment]
Wow, not very good news indeed.

I have some questions/comments for you.

1) By "try without vmware" you mean try on a physical machine, or try to uninstall the vmtools daemon? Because simply put, the second one is feasible, the first one not, as all our systems are virtual machines running on a high availability structure.

2) I'll throw it there: may it be that the problem arises from the non-atomicity of some operations within the lock manager? I don't really know how it works, but as a programmer I know some strange problems can arise from inconsistent assumptions about the states of objects. Please don't read this as a criticims, it's just an idea that comes to mind.

3) It would be great if you can add further logging, I'll do everything I can to help you track down this problem.

**Comment by fperana** [comment]
BTW, FB 3.0.7 worked fine in a very similar machine from a (virtual) hardware point of view, with vmtools installed.
The main differences here are the OS (previously CentOS 7, now Debian 11) and FB 4.

Maybe I can try to switch back to CentOS 7, this would be an expensive step though...

**Comment by AlexPeshkoff** [comment]
On 1/17/23 17:11, F. Perana wrote:
>
> Hey Alex, these last few days we experienced other crashes.
>
> Would you please answer my question about what do you mean by "try 
> without vmware",
>

I've meant running linux with firebird on separate HW, not VM. But I've 
understood from your previous email that it's close to unreal.

> and also if possible please add some more logging in "probably 
> problematic places"?
>

First I want to take a look at your new dumps if possible.
Last week I've fixed one bug that might cause pthread_mutex_lock failed 
or crash, and there may be more reasons for them than one mentioned 
before when shared memory went crazy.


> On my side, I'm going to try FB4 on CentOS 7, but this step will 
> require some work, so I don't know when I'll be able to do it.
>

Probably this can help - but let's begin with new core dumps.

**Comment by fperana** [comment]
> I've meant running linux with firebird on separate HW, not VM. But I've understood from your previous email that it's close to unreal.

Ok, if you meant this, yes, it's impossible, sorry.

> First I want to take a look at your new dumps if possible. Last week I've fixed one bug that might cause pthread_mutex_lock failed or crash, and there may be more reasons for them than one mentioned before when shared memory went crazy.

Here's the last coredump. Actually there were at least two, probably three crashes, but I wasn't here to save them... I don't know if core dumps will append or overwrite...
[core_20230112T160508.001.zip](https://github.com/FirebirdSQL/firebird/files/10436026/core_20230112T160508.001.zip)
[core_20230112T160508.002.zip](https://github.com/FirebirdSQL/firebird/files/10436028/core_20230112T160508.002.zip)
[core_20230112T160508.003.zip](https://github.com/FirebirdSQL/firebird/files/10436029/core_20230112T160508.003.zip)

If you think it can help, I can install your latest version, as far as you'll build one for me that contains only your fixes and nothing more (like the one I'm running now).

**Comment by AlexPeshkoff** [comment]
On 1/17/23 17:43, F. Perana wrote:
>
>     I've meant running linux with firebird on separate HW, not VM. But
>     I've understood from your previous email that it's close to unreal.
>
> Ok, if you meant this, yes, it's impossible, sorry.
>
>     First I want to take a look at your new dumps if possible. Last
>     week I've fixed one bug that might cause pthread_mutex_lock failed
>     or crash, and there may be more reasons for them than one
>     mentioned before when shared memory went crazy.
>
> Here's the last coredump. Actually there were at least two, probably 
> three crashes, but I wasn't here to save them... I don't know if core 
> dumps will append or overwrite...
> core_20230112T160508.001.zip 
> <https://github.com/FirebirdSQL/firebird/files/10436026/core_20230112T160508.001.zip>
> core_20230112T160508.002.zip 
> <https://github.com/FirebirdSQL/firebird/files/10436028/core_20230112T160508.002.zip>
> core_20230112T160508.003.zip 
> <https://github.com/FirebirdSQL/firebird/files/10436029/core_20230112T160508.003.zip>
>
> If you think it can help, I can install your latest version, as far as 
> you'll build one for me that contains only your fixes and nothing more 
> (like the one I'm running now).
>

At least this dump exactly repeats your previous case. I have one idea 
not related with VM / centos / debian, though it's not something 
"strong", just a thought around. Will prepare build for you in a day or two.

Next, do you have linger set for your databases? If not, enter
   ALTER DATABASE SET LINGER TO 600;
for all databases with short attaches lifetime, at least scudonline and 
sicurezza.
This is not absolute fix but should minimize chances of this error to 
happen.

**Comment by fperana** [comment]
> At least this dump exactly repeats your previous case. I have one idea not related with VM / centos / debian, though it's not something "strong", just a thought around. Will prepare build for you in a day or two.

OK thanks. We're now out of the high load period so we'll have to wait until February to see if your patch helps.

>Next, do you have linger set for your databases? If not, enter   ALTER DATABASE SET LINGER TO 600; for all databases with short attaches lifetime, at least scudonline and sicurezza. This is not absolute fix but should minimize chances of this error to happen.

I don't have linger set. But I think that it's better not to set it now, if we want to have a chance to see if your patch works, right?

**Comment by AlexPeshkoff** [comment]
On 1/18/23 14:15, F. Perana wrote:
>
>     At least this dump exactly repeats your previous case. I have one
>     idea not related with VM / centos / debian, though it's not
>     something "strong", just a thought around. Will prepare build for
>     you in a day or two.
>
> OK thanks. We're now out of the high load period so we'll have to wait 
> until February to see if your patch helps.
>
>     Next, do you have linger set for your databases? If not, enter  
>     ALTER DATABASE SET LINGER TO 600; for all databases with short
>     attaches lifetime, at least scudonline and sicurezza. This is not
>     absolute fix but should minimize chances of this error to happen.
>
> I don't have linger set. But I think that it's better not to set it 
> now, if we want to have a chance to see if your patch works, right?
>

If you plan to install patches only in Febraury I recommend to use 
LINGER right now. But solution is certainly your's.

**Comment by AlexPeshkoff** [comment]
BTW, if there are more cores - send them to me.

**Comment by fperana** [comment]
> If you plan to install patches only in Febraury I recommend to use LINGER right now. But solution is certainly your's.

I'll install your build ASAP when it's ready. In the second half of the month we have far less workload on the server, so it's the right time to do such things. Also, we never experienced crashes during this period (for the same reason, low workload).

That said, I don't feel the need to set linger right now, and I think (but I could be wrong here, correct me in that case) that it's better to leave things as they are, so we can see if the new build sorts things out (I mean, touching two things - new build **and** linger - we'll never know wich one solved the problem).

And yes, if a new crash occurs, I'll send you cores.

**Comment by aafemt** [comment]
You can set or unset linger any time, it is not long or dangerous operation.

**Comment by AlexPeshkoff** [comment]
New build
https://drive.google.com/drive/folders/1KXu5S215dN7dsthiAxnStZkttubRFgI-?usp=sharing

**Comment by fperana** [comment]
Here's a new, fresh core dump.

Note that I haven't had time to install your build yet @AlexPeshkoff (I hope to get it up this evening) - and I didn't set LINGER, so the conditions are the same as usual.

[core_20230127T153815.003.zip](https://github.com/FirebirdSQL/firebird/files/10520044/core_20230127T153815.003.zip)
[core_20230127T153815.002.zip](https://github.com/FirebirdSQL/firebird/files/10520042/core_20230127T153815.002.zip)
[core_20230127T153815.001.zip](https://github.com/FirebirdSQL/firebird/files/10520040/core_20230127T153815.001.zip)

**Comment by fperana** [comment]
Another core. And no, I did't had time to install your build yet. I hope tomorrow.

[core_20230130T133135.001.zip](https://github.com/FirebirdSQL/firebird/files/10536655/core_20230130T133135.001.zip)
[core_20230130T133135.002.zip](https://github.com/FirebirdSQL/firebird/files/10536657/core_20230130T133135.002.zip)
[core_20230130T133135.003.zip](https://github.com/FirebirdSQL/firebird/files/10536660/core_20230130T133135.003.zip)

**Comment by AlexPeshkoff** [comment]
Suppose (have strong reason for it) I've found and fixed this bug. Your build is as always:
https://drive.google.com/drive/folders/1KXu5S215dN7dsthiAxnStZkttubRFgI-?usp=sharing

**Comment by fperana** [comment]
Great news! We're now in the middle of our busy period, so I'll try to install the new build this evening.
BTW, still no crashes this month.

**Comment by AlexPeshkoff** [comment]
On 2/8/23 12:22, F. Perana wrote:
>
> Great news! We're now in the middle of our busy period, so I'll try to 
> install the new build this evening.
> BTW, still no crashes this month.
>

Not strange - that's races, random thing, and in our case with very-very 
low possibility - the size of dangerous execution frame was about 10 asm 
instructions.

**Comment by fperana** [comment]
Is the code involved in this race condition new to FB4? I ask this just because with FB3 there were no problems for years.

**Comment by AlexPeshkoff** [comment]
On 2/8/23 12:51, F. Perana wrote:
>
> Is the code involved in this race condition new to FB4?
>

Yes.

> I ask this just because with FB3 there were no problems for years.
>

**Comment by fperana** [comment]
One month later, still no crashes. And it was a very busy month. I think you've sorted it out.

**Comment by fperana** [comment]
I think I spoke too soon. A few days ago the problem reappeared. Probably your patch mitigated it, but it isn't solved.

New core dump follows:

[core_20230321T171237.001.zip](https://github.com/FirebirdSQL/firebird/files/11109313/core_20230321T171237.001.zip)
[core_20230321T171237.002.zip](https://github.com/FirebirdSQL/firebird/files/11109317/core_20230321T171237.002.zip)
[core_20230321T171237.003.zip](https://github.com/FirebirdSQL/firebird/files/11109320/core_20230321T171237.003.zip)

The error found in firebird.log is: Operating system call pthread_mutex_destroy failed. Error code 16

A note: when I installed your patched FB 4.0.2 I didn't have time to also install the corresponding debug files, so I don't know if the coredump is usable. Please let me know... I don't have much time to install them now anyway, maybe next week.

**Comment by AlexPeshkoff** [comment]
Core dumps are usable - debug info is needed not to create dumps but to analyze it.
A reason of an error is unrelated to one we had before, message is almost same cause mutexes are widely used in a code. Please add new ticket for it.
